src/libraries/Hooks.sol

```diff
@@ -151,44 +195,75 @@ library Hooks {
+                        ),
+                        self.hasPermission(AFTER_ADD_LIQUIDITY_RETURNS_DELTA_FLAG)
+                    )
+                );
+            }
+        } else {
+            if (self.hasPermission(AFTER_REMOVE_LIQUIDITY_FLAG)) {
+                hookDelta = BalanceDelta.wrap(
+                    self.callHookWithReturnDelta(
+                        abi.encodeWithSelector(
+                            IHooks.afterRemoveLiquidity.selector, msg.sender, key, params, delta, hookData
+                        ),
+                        self.hasPermission(AFTER_REMOVE_LIQUIDITY_RETURNS_DELTA_FLAG)
+                    )
+                );
+            }
         }
     }
 
     /// @notice calls beforeSwap hook if permissioned and validates return value
     function beforeSwap(IHooks self, PoolKey memory key, IPoolManager.SwapParams memory params, bytes calldata hookData)
         internal
-        noSelfCall(self)
+        returns (int256 amountToSwap, int128 hookDeltaSpecified)
     {
+        amountToSwap = params.amountSpecified;
+        if (msg.sender == address(self)) return (amountToSwap, hookDeltaSpecified);
         if (self.hasPermission(BEFORE_SWAP_FLAG)) {
-            self.callHook(abi.encodeWithSelector(IHooks.beforeSwap.selector, msg.sender, key, params, hookData));
+            hookDeltaSpecified = self.callHookWithReturnDelta(
+                abi.encodeWithSelector(IHooks.beforeSwap.selector, msg.sender, key, params, hookData),
+                self.hasPermission(BEFORE_SWAP_RETURNS_DELTA_FLAG)
+            ).toInt128();
+
+            // Update the swap amount according to the hook's return, and check that the swap type doesnt change (exact input/output)
+            if (hookDeltaSpecified != 0) {
+                bool exactInput = amountToSwap < 0;
+                amountToSwap += hookDeltaSpecified;
+                if (exactInput ? amountToSwap > 0 : amountToSwap < 0) revert HookDeltaExceedsSwapAmount();
+            }
         }
     }
 
     /// @notice calls afterSwap hook if permissioned and validates return value
+    /// @dev if noSelfCall bypasses the call, hookDeltaUnspecified will be 0
     function afterSwap(
         IHooks self,
         PoolKey memory key,
         IPoolManager.SwapParams memory params,
         BalanceDelta delta,
         bytes calldata hookData
-    ) internal noSelfCall(self) {
+    ) internal noSelfCall(self) returns (int128 hookDeltaUnspecified) {
         if (self.hasPermission(AFTER_SWAP_FLAG)) {
-            self.callHook(abi.encodeWithSelector(IHooks.afterSwap.selector, msg.sender, key, params, delta, hookData));
+            hookDeltaUnspecified = self.callHookWithReturnDelta(
+                abi.encodeWithSelector(IHooks.afterSwap.selector, msg.sender, key, params, delta, hookData),
+                self.hasPermission(AFTER_SWAP_RETURNS_DELTA_FLAG)
+            ).toInt128();
```

Comment:
Ok one idea relating to the above comment - it feels like it would be ideal if this function could return `(BalanceDelta hookDelta)` so the top level swap code can look more similar to `afterModifyLiquidity`, i.e. something like 
```
BalanceDelta hookDelta = afterSwap(..)
if (hookDelta > 0) {
  swapperDelta -= hookDelta;
  applyHookDelta;
}
applySwapDelta;
```

maybe would be possible if `afterSwap` had context about the result of `beforeSwap` either via a parameter or tstore?